File: java-kata/src/main/java/com/leetcode/MiddleOfLinkedList.java
package com.leetcode;

import com.leetcode.data.ListNode;

/**
 * <h3><a href="https://leetcode.com/problems/middle-of-the-linked-list/">Middle of the Linked List</a></h3>
 * <p>
 *     Given the <code>head</code> of a singly linked list, return the middle node of the linked list.
 * </p>
 * <p>
 *     If there are two middle nodes, return <strong>the second middle</strong> node.
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li>The number of nodes in the list is in the range <code>[1, 100]</code></li>
 *         <li><code>1 <= Node.val <= 100</code></li>
 *     </ul>
 * </p>
 */
public class MiddleOfLinkedList {

    public ListNode middleNode(ListNode head) {
        var fast = head;
        var slow = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next != null ? fast.next.next : fast.next;
        }
        return slow;
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/MiddleOfLinkedList.kt
package com.leetcode

import com.leetcode.data.ListNode

/**
 * ### [Middle of the Linked List](https://leetcode.com/problems/middle-of-the-linked-list/)
 *
 * Given the `head` of a singly linked list, return the middle node of the linked list.
 *
 * If there are two middle nodes, return **the second middle** node.
 *
 * ###### Constraints:
 *  - The number of nodes in the list is in the range `[1, 100]`
 *  - `1 <= Node.val <= 100`
 */
class MiddleOfLinkedList {

    fun middleNode(head: ListNode?): ListNode? {
        var fast = head
        var slow = head
        while (fast?.next != null) {
            fast = fast.next?.next
            slow = slow?.next
        }
        return slow
    }
}
